File: 00 Basic/Binary Search.py
# https://practice.geeksforgeeks.org/problems/binary-search-1587115620/1?page=1&difficulty[]=-1&curated[]=1&sortBy=submissions

#User function template for Python

class Solution:	
	def binarysearch(self, arr, n, k):
            low = 0
            high = n-1
            while(low<=high):
                mid = (low + high)//2
                if(k == arr[mid]):
                    return mid
                elif(k>arr[mid]):
                    low = mid+1
                else:
                    high = mid-1
            return -1


#{ 
 # Driver Code Starts
#Initial template for Python
if __name__ == '__main__':
    t=int(input())
    for i in range(t):
        n=int(input())
        arr=list(map(int, input().strip().split(' ')))
        k=int(input())
        ob = Solution()
        print (ob.binarysearch(arr, n, k))
# } Driver Code Ends

File: 00 Basic/Binary Search.cpp
// https://practice.geeksforgeeks.org/problems/binary-search-1587115620/1?page=1&difficulty[]=-1&curated[]=1&sortBy=submissions

//{ Driver Code Starts
// Initial template for C++

#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends
// User function template for C++

class Solution {
  public:
    int binarysearch(int arr[], int n, int k) {
        int low=0, mid=0;
        int high=n-1;
        while(low<=high){
            mid = (low + high)/2;
            if(k==arr[mid]){
                return mid;
            }
            else if(k > arr[mid]){
                low = mid+1;
            }
            else{
                high = mid-1;
            }
        }
        return -1;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;

    while (t--) {
        int N;
        cin >> N;
        int arr[N];
        for (int i = 0; i < N; i++) cin >> arr[i];
        int key;
        cin >> key;
        Solution ob;
        int found = ob.binarysearch(arr, N, key);
        cout << found << endl;
    }
}

// } Driver Code Ends